File: scripts/loadConversation.js
#!/usr/bin/env node

/**
 * Load Conversation Utility
 * 
 * Usage: node scripts/loadConversation.js <input-file>
 * 
 * Input file format:
 * ```
 * @user1 @user2
 * 
 * user1: message text
 * user2: reply text
 * user1 (Snap): [Photo description] caption text
 * user2: another message
 * 
 * -- 30 minutes later --
 * 
 * user1: message after time gap
 * ```
 */

const { createClient } = require('@supabase/supabase-js');
const fs = require('fs');
const path = require('path');

// Load environment variables
require('dotenv').config();

const supabaseUrl = process.env.EXPO_PUBLIC_SUPABASE_URL;
const supabaseServiceKey = process.env.SUPABASE_SERVICE_ROLE_KEY || process.env.EXPO_PUBLIC_SUPABASE_ANON_KEY;

if (!supabaseUrl || !supabaseServiceKey) {
  console.error('Error: Missing EXPO_PUBLIC_SUPABASE_URL or SUPABASE_SERVICE_ROLE_KEY');
  process.exit(1);
}

const supabase = createClient(supabaseUrl, supabaseServiceKey);

// Parse time expressions like "30 minutes later", "2 hours later", etc.
function parseTimeGap(timeStr) {
  const match = timeStr.match(/(\d+)\s*(second|minute|hour|day)s?\s*(?:later)?/i);
  if (!match) return 0;
  
  const amount = parseInt(match[1]);
  const unit = match[2].toLowerCase();
  
  const multipliers = {
    second: 1,
    minute: 60,
    hour: 3600,
    day: 86400
  };
  
  return amount * multipliers[unit];
}

// Parse a conversation file
function parseConversationFile(content) {
  const lines = content.split('\n').map(line => line.trim()).filter(line => line);
  
  if (lines.length === 0) {
    throw new Error('Empty file');
  }
  
  // First line should contain @username1 @username2
  const headerMatch = lines[0].match(/@(\w+)\s+@(\w+)/);
  if (!headerMatch) {
    throw new Error('First line must contain two usernames like: @username1 @username2');
  }
  
  const [, username1, username2] = headerMatch;
  const messages = [];
  let currentTimeOffset = 0; // in seconds
  
  for (let i = 1; i < lines.length; i++) {
    const line = lines[i];
    
    // Check for time gap markers
    if (line.startsWith('--') && line.endsWith('--')) {
      const timeGap = parseTimeGap(line);
      currentTimeOffset += timeGap;
      continue;
    }
    
    // Parse message lines
    // Format: username: message text
    // or: username (Type): message text
    const messageMatch = line.match(/^(\w+)(?:\s*\(([^)]+)\))?\s*:\s*(.+)$/);
    if (messageMatch) {
      const [, sender, type, content] = messageMatch;
      
      // Validate sender is one of our users
      if (sender !== username1 && sender !== username2) {
        console.warn(`Warning: Unknown sender "${sender}" on line ${i + 1}`);
        continue;
      }
      
      messages.push({
        sender,
        recipient: sender === username1 ? username2 : username1,
        content,
        type: type ? type.toLowerCase() : 'text',
        timeOffset: currentTimeOffset
      });
      
      // Add small time gaps between messages (30-90 seconds)
      currentTimeOffset += 30 + Math.random() * 60;
    }
  }
  
  return { username1, username2, messages };
}

// Get user IDs from usernames
async function getUserIds(username1, username2) {
  const { data, error } = await supabase
    .from('profiles')
    .select('id, username')
    .in('username', [username1.toLowerCase(), username2.toLowerCase()]);
  
  if (error) {
    throw error;
  }
  
  if (!data || data.length !== 2) {
    throw new Error(`Could not find both users: ${username1}, ${username2}`);
  }
  
  const userMap = {};
  data.forEach(user => {
    userMap[user.username] = user.id;
  });
  
  return {
    [username1.toLowerCase()]: userMap[username1.toLowerCase()],
    [username2.toLowerCase()]: userMap[username2.toLowerCase()]
  };
}

// Generate room ID for direct messages
function getRoomId(userId1, userId2) {
  return `dm_${[userId1, userId2].sort().join('_')}`;
}

// Load conversation into database
async function loadConversation(filePath) {
  try {
    // Read and parse file
    const content = fs.readFileSync(filePath, 'utf8');
    const { username1, username2, messages } = parseConversationFile(content);
    
    console.log(`Loading conversation between @${username1} and @${username2}`);
    console.log(`Found ${messages.length} messages`);
    
    // Get user IDs
    const userIds = await getUserIds(username1, username2);
    const userId1 = userIds[username1.toLowerCase()];
    const userId2 = userIds[username2.toLowerCase()];
    
    if (!userId1 || !userId2) {
      throw new Error('Could not find user IDs');
    }
    
    // Generate room ID
    const roomId = getRoomId(userId1, userId2);
    console.log(`Room ID: ${roomId}`);
    
    // Clear existing messages in this room (optional)
    // Note: This requires either service role key or DELETE RLS policy
    if (supabaseServiceKey && supabaseServiceKey !== process.env.EXPO_PUBLIC_SUPABASE_ANON_KEY) {
      console.log('Clearing existing messages...');
      const { error: deleteError } = await supabase
        .from('messages')
        .delete()
        .eq('room_id', roomId);
      
      if (deleteError) {
        console.warn('Warning: Could not clear existing messages:', deleteError.message);
        console.warn('Continuing with insertion (may create duplicates)...');
      }
    } else {
      console.log('Using anon key - skipping message deletion to avoid permission errors');
      console.log('Note: This may create duplicate messages if conversation already exists');
    }
    
    // Prepare messages for insertion
    const now = new Date();
    const messagesToInsert = messages.map(msg => {
      const senderId = userIds[msg.sender.toLowerCase()];
      const recipientId = userIds[msg.recipient.toLowerCase()];
      const createdAt = new Date(now.getTime() - (msg.timeOffset * 1000));
      
      return {
        room_id: roomId,
        sender_id: senderId,
        recipient_id: recipientId,
        content: msg.content,
        type: msg.type === 'snap' || msg.type === 'photo' || msg.type === 'video' ? 'text' : msg.type,
        created_at: createdAt.toISOString()
      };
    });
    
    // Insert messages in batches of 50
    const batchSize = 50;
    let inserted = 0;
    
    for (let i = 0; i < messagesToInsert.length; i += batchSize) {
      const batch = messagesToInsert.slice(i, i + batchSize);
      
      const { error: insertError } = await supabase
        .from('messages')
        .insert(batch);
      
      if (insertError) {
        throw insertError;
      }
      
      inserted += batch.length;
      console.log(`Inserted ${inserted}/${messagesToInsert.length} messages...`);
    }
    
    console.log(`✅ Successfully loaded ${messagesToInsert.length} messages!`);
    console.log(`\nConversation loaded between @${username1} and @${username2}`);
    
  } catch (error) {
    console.error('Error loading conversation:', error.message);
    process.exit(1);
  }
}

// Main execution
const filePath = process.argv[2];

if (!filePath) {
  console.log(`
Usage: node scripts/loadConversation.js <input-file>

Example input file format:

@username1 @username2

username1: Hey, how's it going?
username2: Pretty good! Just working on some code.
username1 (Snap): [Photo of coffee] Need caffeine!
username2: Haha I feel that

-- 30 minutes later --

username1: Want to grab lunch?
username2: Sure! Where were you thinking?
`);
  process.exit(1);
}

if (!fs.existsSync(filePath)) {
  console.error(`Error: File not found: ${filePath}`);
  process.exit(1);
}

// Load the conversation
loadConversation(filePath);

File: scripts/loadConversation.py
#!/usr/bin/env python3

"""
Load Conversation Utility

Usage: python scripts/loadConversation.py <input-file>

Input file format:
```
@user1 @user2

user1: message text
user2: reply text
user1 (Snap): [Photo description] caption text
user2: another message

-- 30 minutes later --

user1: message after time gap
```
"""

import os
import sys
import re
import json
from datetime import datetime, timedelta
from typing import List, Dict, Tuple
import requests
from dotenv import load_dotenv

# Load environment variables
load_dotenv()

SUPABASE_URL = os.getenv('EXPO_PUBLIC_SUPABASE_URL')
SUPABASE_KEY = os.getenv('SUPABASE_SERVICE_ROLE_KEY') or os.getenv('EXPO_PUBLIC_SUPABASE_ANON_KEY')

if not SUPABASE_URL or not SUPABASE_KEY:
    print("Error: Missing EXPO_PUBLIC_SUPABASE_URL or SUPABASE_SERVICE_ROLE_KEY")
    sys.exit(1)

# Supabase API headers
HEADERS = {
    'apikey': SUPABASE_KEY,
    'Authorization': f'Bearer {SUPABASE_KEY}',
    'Content-Type': 'application/json'
}

def parse_time_gap(time_str: str) -> int:
    """Parse time expressions like '30 minutes later', '2 hours later', etc."""
    match = re.search(r'(\d+)\s*(second|minute|hour|day)s?\s*(?:later)?', time_str, re.IGNORECASE)
    if not match:
        return 0
    
    amount = int(match.group(1))
    unit = match.group(2).lower()
    
    multipliers = {
        'second': 1,
        'minute': 60,
        'hour': 3600,
        'day': 86400
    }
    
    return amount * multipliers.get(unit, 0)

def parse_conversation_file(content: str) -> Dict:
    """Parse a conversation file and return structured data."""
    lines = [line.strip() for line in content.split('\n') if line.strip()]
    
    if not lines:
        raise ValueError("Empty file")
    
    # First line should contain @username1 @username2
    header_match = re.match(r'@(\w+)\s+@(\w+)', lines[0])
    if not header_match:
        raise ValueError("First line must contain two usernames like: @username1 @username2")
    
    username1, username2 = header_match.groups()
    messages = []
    current_time_offset = 0  # in seconds
    
    for i, line in enumerate(lines[1:], 1):
        # Check for time gap markers
        if line.startswith('--') and line.endswith('--'):
            time_gap = parse_time_gap(line)
            current_time_offset += time_gap
            continue
        
        # Parse message lines
        # Format: username: message text
        # or: username (Type): message text
        message_match = re.match(r'^(\w+)(?:\s*\(([^)]+)\))?\s*:\s*(.+)$', line)
        if message_match:
            sender, msg_type, content = message_match.groups()
            
            # Validate sender is one of our users
            if sender not in [username1, username2]:
                print(f"Warning: Unknown sender '{sender}' on line {i + 1}")
                continue
            
            recipient = username2 if sender == username1 else username1
            
            messages.append({
                'sender': sender,
                'recipient': recipient,
                'content': content,
                'type': msg_type.lower() if msg_type else 'text',
                'timeOffset': current_time_offset
            })
            
            # Add small time gaps between messages (30-90 seconds)
            import random
            current_time_offset += random.randint(30, 90)
    
    return {
        'username1': username1,
        'username2': username2,
        'messages': messages
    }

def get_user_ids(username1: str, username2: str) -> Dict[str, str]:
    """Get user IDs from usernames."""
    url = f"{SUPABASE_URL}/rest/v1/profiles"
    params = {
        'select': 'id,username',
        'username': f'in.({username1.lower()},{username2.lower()})'
    }
    
    response = requests.get(url, headers=HEADERS, params=params)
    if response.status_code != 200:
        raise Exception(f"Failed to fetch users: {response.text}")
    
    data = response.json()
    if len(data) != 2:
        raise Exception(f"Could not find both users: {username1}, {username2}")
    
    user_map = {user['username']: user['id'] for user in data}
    return {
        username1.lower(): user_map.get(username1.lower()),
        username2.lower(): user_map.get(username2.lower())
    }

def get_room_id(user_id1: str, user_id2: str) -> str:
    """Generate room ID for direct messages."""
    return f"dm_{sorted([user_id1, user_id2])[0]}_{sorted([user_id1, user_id2])[1]}"

def clear_existing_messages(room_id: str) -> bool:
    """Clear existing messages in a room. Returns True if successful."""
    # Check if we're using service role key (it's longer than anon key)
    is_service_role = SUPABASE_KEY and len(SUPABASE_KEY) > 40 and not SUPABASE_KEY.startswith('eyJ')
    
    if not is_service_role:
        print("Using anon key - skipping message deletion to avoid permission errors")
        print("Note: This may create duplicate messages if conversation already exists")
        return False
    
    url = f"{SUPABASE_URL}/rest/v1/messages"
    params = {'room_id': f'eq.{room_id}'}
    
    response = requests.delete(url, headers=HEADERS, params=params)
    if response.status_code not in [200, 204]:
        print(f"Warning: Could not clear existing messages: {response.text}")
        print("Continuing with insertion (may create duplicates)...")
        return False
    return True

def insert_messages(messages: List[Dict]):
    """Insert messages in batches."""
    url = f"{SUPABASE_URL}/rest/v1/messages"
    batch_size = 50
    inserted = 0
    
    for i in range(0, len(messages), batch_size):
        batch = messages[i:i + batch_size]
        
        response = requests.post(url, headers=HEADERS, json=batch)
        if response.status_code not in [200, 201]:
            raise Exception(f"Failed to insert messages: {response.text}")
        
        inserted += len(batch)
        print(f"Inserted {inserted}/{len(messages)} messages...")

def load_conversation(file_path: str):
    """Main function to load a conversation from a file."""
    try:
        # Read and parse file
        with open(file_path, 'r', encoding='utf-8') as f:
            content = f.read()
        
        parsed = parse_conversation_file(content)
        username1 = parsed['username1']
        username2 = parsed['username2']
        messages = parsed['messages']
        
        print(f"Loading conversation between @{username1} and @{username2}")
        print(f"Found {len(messages)} messages")
        
        # Get user IDs
        user_ids = get_user_ids(username1, username2)
        user_id1 = user_ids[username1.lower()]
        user_id2 = user_ids[username2.lower()]
        
        if not user_id1 or not user_id2:
            raise Exception("Could not find user IDs")
        
        # Generate room ID
        room_id = get_room_id(user_id1, user_id2)
        print(f"Room ID: {room_id}")
        
        # Clear existing messages
        print("Clearing existing messages...")
        clear_existing_messages(room_id)
        
        # Prepare messages for insertion
        now = datetime.now()
        messages_to_insert = []
        
        for msg in messages:
            sender_id = user_ids[msg['sender'].lower()]
            recipient_id = user_ids[msg['recipient'].lower()]
            created_at = now - timedelta(seconds=msg['timeOffset'])
            
            # Handle different message types
            msg_type = 'text'
            if msg['type'] in ['snap', 'photo', 'video']:
                # For now, we'll represent media as text descriptions
                msg_type = 'text'
            
            messages_to_insert.append({
                'room_id': room_id,
                'sender_id': sender_id,
                'recipient_id': recipient_id,
                'content': msg['content'],
                'type': msg_type,
                'created_at': created_at.isoformat()
            })
        
        # Insert messages
        insert_messages(messages_to_insert)
        
        print(f"✅ Successfully loaded {len(messages_to_insert)} messages!")
        print(f"\nConversation loaded between @{username1} and @{username2}")
        
    except Exception as e:
        print(f"Error loading conversation: {e}")
        sys.exit(1)

def show_help():
    """Show usage help."""
    print("""
Load Conversation Utility

Usage: python scripts/loadConversation.py <input-file>

Example input file (save as conversation.txt):

@alice @bob

alice: Hey! How's your day going?
bob: Pretty good! Just finished a big project
alice: Nice! Want to celebrate?
bob: Absolutely! What did you have in mind?

-- 20 minutes later --

alice: How about that new sushi place?
bob: Perfect! I've been wanting to try it
alice (Snap): [Photo of menu] The menu looks amazing!
bob: Wow, those rolls look incredible

-- 1 hour later --

alice: That was delicious!
bob: Best sushi I've had in ages
alice: We should do this more often
bob: Definitely! Same time next week?
alice: It's a date!

Special formats:
- Time gaps: -- X minutes/hours/days later --
- Media messages: username (Snap): [description] caption
- Or: username (Photo): [description]
- Or: username (Video): [description]
""")

if __name__ == '__main__':
    if len(sys.argv) < 2:
        show_help()
        sys.exit(1)
    
    file_path = sys.argv[1]
    
    if not os.path.exists(file_path):
        print(f"Error: File not found: {file_path}")
        sys.exit(1)
    
    load_conversation(file_path)